### PR TITLE
Feature/batching

### PR DESCRIPTION
**Pull request by slvnwhrl**
This PR merges the batching branch containing changes to allow "proper" matching for the neural transducer.

Major changes in this PR:
- Batched training: Training can now be performed with batches.
- Batched greedy decoding: Greedy decoding can now be performed in batches as well.
- Gradient accumulation: Is still possible and can be combined with batches.
- Unit tests are adapted to the changes.

What's more, merging this branch into development should then then allow to merge #2  with less complications.

### CODE REVIEW

**Comment by slvnwhrl** [comment]
@peter-makarov I would like to push these changes into the development branch such that I can then push these encoder-related changes (see #2). This will allow to combine batching with the Transformer-based encoder. Let me know if you see any problems with my recent code changes.

File: trans/train.py [hunk 1]
```diff
@@ -147,30 +200,28 @@ def main(args: argparse.Namespace):
...
+                losses.sum().backward()
+                if j % args.grad_accumulation == 0:
+                    optimizer.step()
+                    transducer_.zero_grad()
                 if j > 0 and j % 100 == 0:
                     logging.info("\t\t...%d batches", j)
             logging.info("\t\t...%d batches", j + 1)
 
+        # avg. loss per sample
         avg_loss = train_loss / len(training_data_loader)
```
Comment (peter-makarov) [inline trans/train.py:225]: Shouldn't `train_loss` be divided by the the number of batches rather than the number of samples?

File: trans/train.py [hunk 2]
```diff
@@ -251,14 +301,13 @@ def main(args: argparse.Namespace):
...
 if __name__ == "__main__":
-
     logging.basicConfig(level="INFO", format="%(levelname)s: %(message)s")
 
     parser = argparse.ArgumentParser(
         description="Train a g2p neural transducer.")
 
     '''parser.add_argument("--dynet-seed", type=int, required=True,
-                        help="DyNET random seed.")''' # TODO: how to replace if at all?
+                        help="DyNET random seed.")'''  # TODO: how to replace if at all?
```
Comment (peter-makarov) [inline trans/train.py:310]: We should probably eventually replace this with `torch` random seed.
https://pytorch.org/docs/stable/notes/randomness.html

File: trans/train.py (hunk 1 above)
Comment (slvnwhrl) [reply trans/train.py:225]: Yes, `len(training_data_loader)` is equal to the number of batches.

File: trans/train.py (hunk 2 above)
Comment (slvnwhrl) [reply trans/train.py:310]: Absolutely, this is part of the changes of PR #2.

**Comment by peter-makarov** [comment]
Sorry I didn't have the time to finish this yesterday :/ 

I believe the code is sound. One thing that we could look into: Currently, you do alignment via indexing. Maybe what we could do going forward is replace it with multiplication with an alignment tensor (a tensor of 1s and 0s). I know this practice from other projects. Alignment via indexing may have a computational overhead and is arguably less readable.

Other than this, great job! :1st_place_medal:

**Comment by peter-makarov** [comment]
Sorry I couldn't review it fast enough! Still, I will try to finish going
through the code and leave comments.

Thanks a lot, great stuff!
